flang/lib/Optimizer/Transforms/OMPMapInfoFinalization.cpp

```diff
@@ -0,0 +1,262 @@
+  // guarantee the parent will be mapped in its totality, partial
+  // mapping is common.
+  //
+  // For example:
+  //    map(tofrom: x%y)
+  //
+  // Will generate a mapping for "x" (the parent) and "y" (the member),
+  // the parent "x" will not be mapped, only the member "y" will,
+  // however, we must have the parent as a BlockArg and MapOperand in
+  // these cases to maintain the correct uses within the region and
+  // it helps to track that the member is part of a larger object.
+  //
+  // In the case of:
+  //    map(tofrom: x%y, x%z)
+  //
+  // The parent member becomes more critical, as we perform a partial
+  // structure mapping, where we link the mapping of x and y together
+  // via the parent (at a kernel argument level in LLVM IR not just
+  // MLIR, important to maintain similarity to Clang and for the runtime
+  // to do the correct thing), however, we still do not map the structure
+  // in its totality, we do however, generate an un-sized "binding"
+  // map entry for it.
+  //
+  // In the case of:
+  //    map(tofrom: x, x%y, x%z)
+  //
+  // We do actually map the entirety of "x", so the explicit
+  // mapping of x%y, x%z becomes unneccesary. It also doesn't
+  // quite make sense to write this from a Fortran OpenMP
+  // perspective (although it is legal), as even if the members were
+  // allocatables or pointers, we are mandated by the specification
+  // to map these (and any recursive components) in their entirety,
+  // which is different to the C++ equivelant, which requires
+  // explicit mapping of these segments.
+  void addImplicitMembersToTarget(mlir::omp::MapInfoOp op,
+                                  fir::FirOpBuilder &builder,
+                                  mlir::Operation *target) {
+    if (auto mapClauseOwner =
+            llvm::dyn_cast<mlir::omp::MapClauseOwningOpInterface>(target)) {
+      llvm::SmallVector<mlir::Value> newMapOps;
+      mlir::OperandRange mapOperandsArr = mapClauseOwner.getMapOperands();
+
+      for (size_t i = 0; i < mapOperandsArr.size(); ++i) {
+        if (mapOperandsArr[i] == op) {
+          // Push member maps
+          for (auto member : op.getMembers()) {
+            newMapOps.push_back(member);
+            // for TargetOp's which have IsolatedFromAbove we must align the
+            // new additional map operand with an appropriate BlockArgument,
+            // as the printing and later processing currently requires a 1:1
+            // mapping of BlockArgs to MapInfoOp's at the same placement in
+            // each array (BlockArgs and MapOperands).
+            if (auto targetOp = llvm::dyn_cast<mlir::omp::TargetOp>(target))
+              targetOp.getRegion().insertArgument(i, member.getType(),
+                                                  builder.getUnknownLoc());
+          }
+        }
+        newMapOps.push_back(mapOperandsArr[i]);
+      }
+      mapClauseOwner.getMapOperandsMutable().assign(newMapOps);
```

Comment:
Wouldn't that overwrite other map operands that are neither the record or its members? What happens if you are mapping from 2 different records for example?

Reply:
It shouldn't at least from my understanding (which may be flawed), in the above case, we loop over the map operands of the owner (e.g. TargetOp) and generate a new vector of map operands to replace the previous, for every MapInfoOp in the map operands that is not the MapInfoOp we found in our walk and are currently processing we insert it into the new list at the same point, for the cases where it's the MapInfoOp we are processing we check if it has any members and directly insert them in-front of there parent on the list, and create a corresponding BlockArg if necessary, this new vector then replaces the old with the assign.

It's a little convoluted looking but unfortunately some of the Mutable interfaces you get from various MLIR types don't provide the best set of functionality, e.g. I don't believe in this particular case we can insert into the existing list at a particular index, which would be more concise, at least from what I've found, but perhaps I am missing some key thing!

I could add a test case for this in derived-type-map.f90 to double check all works as expected, there's a few runtime tests in one of the other related PRs that do this, but I haven't actually added the case when checking the IR itself.

This is very similar to what occurs for descriptors currently, so it's mimicked, however, with some hindsight it may have been better to walk the module for each of the map holding Target operations (TargetOp/EnterOp etc.), as opposed to the MapInfoOp, some of the code repetition required for it is likely alleviatable by templates, however, the main downside to that approach is that whenever we add a new map oeprand carrying operation . The upside is the pass would be a lot quicker, there's a lot of repetition with the current approach that's not always necessary. I'd leave a change like that for a follow up refactoring PR though personally as it touches the current descriptor mapping as well.

Reply:
Thanks for the clarification, makes sense.